module/apmstreadwayamqp/pub.go

```diff
@@ -0,0 +1,72 @@
+package apmstreadwayamqp
+
+import (
+	"context"
+	"github.com/streadway/amqp"
+	"go.elastic.co/apm/v2"
+)
+
+// Publisher is the interface returned by ContextConn.
+//
+// Publisher's Publish method reports spans using the bound context.
+type Publisher interface {
+
+	// WithContext returns a shallow copy of the Publisher with
+	// its context changed to ctx.
+	//
+	// To report publishing as spans, ctx must contain a transaction or span.
+	WithContext(ctx context.Context) Publisher
+
+	// Publish publishes a message and returns an error in encountered.
+	//
+	// The started span is stored in the ctx.
+	Publish(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error
+}
+
+// Wrap wraps amqp.Channel such that its Publish method calls streadway/amqp.Publish.
+func Wrap(ch *amqp.Channel) Publisher {
+	return msgPublisher{ch: ch, ctx: context.Background()}
+}
+
+type msgPublisher struct {
+	ch  *amqp.Channel
+	ctx context.Context
+}
+
+func (c msgPublisher) WithContext(ctx context.Context) Publisher {
+	c.ctx = ctx
+	return c
+}
+
+func (c msgPublisher) Publish(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
+	return pub(c.ctx, c.ch, exchange, key, mandatory, immediate, msg)
+}
+
+func pub(ctx context.Context, ch *amqp.Channel, exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
```

Comment:
Instead of creating an interface with a bound context, can we expose this function as a package-level function? i.e.

```go
// Publish calls ch.Publish.
//
// If ctx contains an apm.Transaction, Publish will create a span and inject trace context into msg.Headers.
//
// NOTE: ctx is not used for cancellation.
func Publish(ctx context.Context, ch *amqp.Channel, exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error
```